lib/services/local/usagereporter.go

```diff
@@ -194,17 +212,33 @@ func (r *UsageReporter) runSubmit(ctx context.Context) {
 		case batch := <-r.submissionQueue:
 			t0 := time.Now()
 
-			if err := r.submit(r, batch); err != nil {
-				r.Warnf("Failed to submit batch of %d usage events: %v", len(batch), err)
-
+			if failed, err := r.submit(r, batch); err != nil {
+				r.WithField("batch_size", len(batch)).Warnf("failed to submit batch of usage events: %v", err)
 				usageBatchesFailed.Inc()
 
+				var dropped []*SubmittedEvent
+				var resubmit []*SubmittedEvent
+				for _, e := range failed {
+					e.retriesRemaining -= 1
+
+					if e.retriesRemaining > 0 {
+						resubmit = append(resubmit, e)
+					} else {
+						dropped = append(dropped, e)
```

Comment:
Can `dropped` just be an integer counter? It doesn't seem like you do anything with the events you add to the slice, so it's just unnecessary allocations.

Reply:
Good catch, we don't even need that since we can just calculate the value.